### PR TITLE
Switch branch cache from in-memory cache to BigTable

### PR DESCRIPTION
**Pull request by shifucun**
This is a big PR.. but it includes all necessary changes to completely swap memory cache to branch BT.

- Use a store to wrap the base and branch Bigtable clients. Use a RWLock to handle branch table update and read.
- Many APIs are simplified as the dual bigtable read is handled by the bigTableReadRowsParallel function.

---
This change is [Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444)

### CODE REVIEW

**Comment by pradh** [comment]
Nice... This is a big change :) I'll take a careful look first thing next week...

**Comment by pradh** [comment]
Might be good to get another pair of eyes, hanlu?

**Comment by shifucun** [comment]
> Might be good to get another pair of eyes, hanlu?

Added more people :)

File: deploy/base/deployment.yaml [hunk 1]
```diff
@@ -77,11 +78,11 @@ spec:
             exec:
               command: ["/bin/grpc_health_probe", "-addr=:12345"]
             periodSeconds: 5
-          # livenessProbe:
-          #   exec:
-          #     command: ["/bin/grpc_health_probe", "-addr=:12345"]
-          #   periodSeconds: 5
-          #   initialDelaySeconds: 10
+          livenessProbe:
```
[the remark this answers is not in the document]
Comment (shifucun) [reply deploy/base/deployment.yaml:81]: This is a correct. line 61 and 76 was hidden in Github, so this looks odd

File: internal/server/model.go [hunk 2]
```diff
@@ -98,11 +98,13 @@ type PlaceStatsVar struct {
...
-	InArcInfo  map[string][]translator.InArcInfo
-	SubTypeMap map[string]string
-	Bq         string
+	Mappings       []*base.Mapping
+	OutArcInfo     map[string]map[string][]translator.OutArcInfo
+	InArcInfo      map[string][]translator.InArcInfo
+	SubTypeMap     map[string]string
+	Bq             string
+	BtProject      string
+	branchInstance string
```
[the remark this answers is not in the document]
Comment (shifucun) [reply internal/server/model.go]: Yep, updated

**Review by beets — APPROVED** [review]
thx!

**Review by pradh — CHANGES_REQUESTED** [review]
Nice simplification! Main comments around concurrent updates to BT client...

> Reviewed 33 of 43 files at r1, 12 of 13 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444#-:-MRREBeqCbvKj8jb9Dud:beowvwh)* status: 47 of 48 files reviewed, 13 unresolved discussions (waiting on @beets, @hanlu09205, and @shifucun)

___
*[cmd/main.go, line 42 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9Mb67UX7Ts_uXEOI:-MRR9Mb67UX7Ts_uXEOJ:bpprlbt) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L42)):*
> ```
> base_instance
> ```
base_bt_instance ?
___
*[cmd/main.go, line 44 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9Pkx9XZ6rlCaJr4L:-MRR9Pkx9XZ6rlCaJr4M:b18oglf) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L44)):*
> ```
> branch_instance
> ```
branch_bt_instance?
___
*[cmd/main.go, line 98 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR7LtqCHGQ1svaWdgr:-MRR7LtqCHGQ1svaWdgs:bg5etb) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L98)):*
> ```
> btClient = branchClient
> ```
Comment on reason why btClient is branchClient? Or that it doesn't matter which one it is.
___
*[internal/server/model.go, line 107 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9mIqEbqB7z0JiTjU:-MRR9mIqEbqB7z0JiTjV:b-12rvle) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/model.go#L107)):*
> ```Go
> 	Bq             string
> 	BtProject      string
> 	BranchInstance string
> ```

BranchBTInstance
___
*[internal/server/populations.go, line 63 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRBo8iE59I0nyWphs8:-MRRBo8iE59I0nyWphs9:b1kc7y8) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/populations.go#L63)):*
> ```
> 0
> ```
here too using index might be nicer...
___
*[internal/server/property_label.go, line 44 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRC7Dd0lb5ImL-ozoB:-MRRC7Dd0lb5ImL-ozoC:b-34tf1w) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/property_label.go#L44)):*
> ```
> // Property Labels cache only read from the base cache.
> ```

unrelated to this PR: Is it a TODO to merge labels across branch and base?
___
*[internal/server/server.go, line 49 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRBcCo8naUG6wTV7qh:-MRRBcCo8naUG6wTV7qi:b-74sg3t) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L49)):*
> ```
> tables
> ```
nit: btTables ?
___
*[internal/server/server.go, line 52 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRA9fh1kFNZuA_OJxK:-MRRA9fh1kFNZuA_OJxL:b-5r9pdq) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L52)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> func (s *Server) tables() []*bigtable.Table {
> 	s.RLock()
> 	defer s.RUnlock()
> 	return s.btTables
> }
> ```
</details>


To confirm, are we making a deep-copy here? Otherwise, it is not safe because updateBranchTable could happen?
___
*[internal/server/server.go, line 64 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRB8Ru03o0yiMUBDRA:-MRRB8Ru03o0yiMUBDRB:b-h8kxek) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L64)):*
> ```
> s.btClient.Close()
> ```

Can there be accesses to btTables[0] created from previous version? If so, is it safe to close the client while accesses are going on? If not we might need some sort of ref counting...
___
*[internal/server/server.go, line 66 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR8CbU6gYUph7JOzDo:-MRR8CbU6gYUph7JOzDp:b-wuet26) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L66)):*
> ```
> s.btTables[0] = branchTable
> ```

Feels like a hidden assumption. We could consider using kBranchBTIndex and kBaseBTIndex and use those indexes in main.go and here...
___
*[test/integration/setup.go, line 146 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRDw_M3hNscEiyy7uS:-MRRDw_M3hNscEiyy7uT:b-abs222) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/test/integration/setup.go#L146)):*
> ```
> server.SetupBigtable(
> ```
For my understanding what does this use (mock or real instance)?

**Review by shifucun — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444)* status: 47 of 48 files reviewed, 13 unresolved discussions (waiting on @beets, @hanlu09205, and @pradh)

___
*[cmd/main.go, line 42 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9Mb67UX7Ts_uXEOI:-MRSAkfk4UiPq7mfbvFs:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L42)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> base_instance
> ```
base_bt_instance ?
</blockquote></details>

Done.
___
*[cmd/main.go, line 44 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9Pkx9XZ6rlCaJr4L:-MRSAk0t8Fn2eF92jIej:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L44)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> branch_instance
> ```
branch_bt_instance?
</blockquote></details>

Done.
___
*[cmd/main.go, line 98 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR7LtqCHGQ1svaWdgr:-MRSBIp-Cg2MbpQw9een:b-3wfmjb) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/cmd/main.go#L98)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> btClient = branchClient
> ```
Comment on reason why btClient is branchClient? Or that it doesn't matter which one it is.
</blockquote></details>

changed to branchBtClient and comment on the need to update it in branch cache read
___
*[internal/server/model.go, line 107 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR9mIqEbqB7z0JiTjU:-MRSBemeAevfSQ82OUOj:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/model.go#L107)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

BranchBTInstance
</blockquote></details>

Done.
___
*[internal/server/populations.go, line 63 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRBo8iE59I0nyWphs8:-MRSWa6U6PVXtPiO6-ST:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/populations.go#L63)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> 0
> ```
here too using index might be nicer...
</blockquote></details>

Done.
___
*[internal/server/property_label.go, line 44 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRC7Dd0lb5ImL-ozoB:-MRSC-ChCQMSwmJEQSlZ:bqp6f1w) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/property_label.go#L44)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> // Property Labels cache only read from the base cache.
> ```

unrelated to this PR: Is it a TODO to merge labels across branch and base?
</blockquote></details>

Here uses a simplification that branch cache merge only involves stats data but not schema
___
*[internal/server/server.go, line 49 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRBcCo8naUG6wTV7qh:-MRSDqii30k1ib8LgvfE:b-ap8na7) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L49)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> tables
> ```
nit: btTables ?
</blockquote></details>

now directly access the btTables field
___
*[internal/server/server.go, line 52 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRA9fh1kFNZuA_OJxK:-MRSE5dt2HdTk0YVUI-f:bkp1vrq) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L52)):*
> updateBranchTable

as discussed, update the pointer and let the old client leak/close on server restart
___
*[internal/server/server.go, line 64 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRB8Ru03o0yiMUBDRA:-MRSEImK2iPjlMSFsY8o:b-rvjcjq) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L64)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> s.btClient.Close()
> ```

Can there be accesses to btTables[0] created from previous version? If so, is it safe to close the client while accesses are going on? If not we might need some sort of ref counting...
</blockquote></details>

Ideally no client should be able to call 
___
*[internal/server/server.go, line 66 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRR8CbU6gYUph7JOzDo:-MRSVcQ775f9L3TNvqy5:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/internal/server/server.go#L66)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> s.btTables[0] = branchTable
> ```

Feels like a hidden assumption. We could consider using kBranchBTIndex and kBaseBTIndex and use those indexes in main.go and here...
</blockquote></details>

Done.
___
*[test/integration/setup.go, line 146 at r3](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRRDw_M3hNscEiyy7uS:-MRSCTxeDwKfsowJYUZg:ba97uj2) ([raw file](https://github.com/datacommonsorg/mixer/blob/2fcbf536a4fb3b3eb4a4ff0a3fd494fd663d5e15/test/integration/setup.go#L146)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> server.SetupBigtable(
> ```
For my understanding what does this use (mock or real instance)?
</blockquote></details>

This is a in-memory bigtable instance with mock data

**Review by shifucun — COMMENTED** [review]
Now just update the branch bt client during reload and let the old client leak/close on server exit.

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444#-:-MRSWsmq92BHZ2raCvYE:b358pir)* status: 47 of 48 files reviewed, 13 unresolved discussions (waiting on @beets, @hanlu09205, and @pradh)

**Review by pradh — APPROVED** [review]
Thanks for making the changes!

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444#-:-MRXJQfOEOS2vZk6F4_Y:bl137vd)* status: 27 of 53 files reviewed, 6 unresolved discussions (waiting on @beets, @hanlu09205, @pradh, and @shifucun)

___
*[internal/server/populations.go, line 59 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXGg_S0IfQwv7EALLd:-MRXGg_S0IfQwv7EALLe:b-4xjz8x) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/server/populations.go#L59)):*
> ```
> s.store.BranchBt()
> ```

Check BranchBt() and BaseBt() for null before calling read-row on them? (the removed check at the function top)
___
*[internal/server/populations.go, line 116 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXGsXU6ZwnrY4J0nYL:-MRXGsXU6ZwnrY4J0nYM:b-v2diui) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/server/populations.go#L116)):*
> ```
> if s.store.BaseBt() == nil && s.store.BranchBt() == nil {
> ```
If either is specified we continue, so if branch is null, s.store.BranchBt().ReadRow() below can crash?
___
*[internal/store/store.go, line 15 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXJLdv3DICBCgQu8pQ:-MRXIrit5co4gbtiCTE1:b-ycssq4) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/store/store.go#L15)):*
> ```Go
> // limitations under the License.
> 
> package store
> ```

This looks cleaner!
___
*[internal/store/store.go, line 29 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXIfZpBhgirSHanojZ:-MRXIfZpBhgirSHanoj_:b-h58bbc) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/store/store.go#L29)):*
> ```
> sync.RWMutex
> ```

nit: for clarity, use a variable branchLock ?

**Review by shifucun — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444)* status: 27 of 53 files reviewed, 5 unresolved discussions (waiting on @beets, @hanlu09205, and @pradh)

___
*[internal/server/populations.go, line 59 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXGg_S0IfQwv7EALLd:-MRXK2bp-Py410_rM__5:b-896fix) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/server/populations.go#L59)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> s.store.BranchBt()
> ```

Check BranchBt() and BaseBt() for null before calling read-row on them? (the removed check at the function top)
</blockquote></details>

Done.
___
*[internal/server/populations.go, line 116 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXGsXU6ZwnrY4J0nYL:-MRXKdmkBw63MrVI10Qm:b-lm3jah) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/server/populations.go#L116)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> if s.store.BaseBt() == nil && s.store.BranchBt() == nil {
> ```
If either is specified we continue, so if branch is null, s.store.BranchBt().ReadRow() below can crash?
</blockquote></details>

good catch, should both be non-nil
___
*[internal/store/store.go, line 15 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXJLdv3DICBCgQu8pQ:-MRXLJ46B_qSAbmvmTPJ:b-pl14a5) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/store/store.go#L15)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

This looks cleaner!
</blockquote></details>


Thanks!
___
*[internal/store/store.go, line 29 at r6](https://reviewable.io/reviews/datacommonsorg/mixer/444#-MRXIfZpBhgirSHanojZ:-MRXLKys0TR-e_HpqvsR:bb6un1j) ([raw file](https://github.com/datacommonsorg/mixer/blob/3f05edc6cbe12dee01cf1c2f23f8cdadae6ad572/internal/store/store.go#L29)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> sync.RWMutex
> ```

nit: for clarity, use a variable branchLock ?
</blockquote></details>

Done

**Review by shifucun — COMMENTED** [review]
Thanks for review!

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/mixer/444#-:-MRXLXyfBwJQAyogDqXg:boi3pfu)* status: 27 of 53 files reviewed, 5 unresolved discussions (waiting on @beets, @hanlu09205, and @pradh)

**Pull request merged**